presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java

```diff
@@ -0,0 +1,542 @@
+        }
+        catch (Exception e) {
+            throw new PrestoException(COMPILER_ERROR, e.getCause());
+        }
+
+        return () -> {
+            try {
+                return projectionClass.newInstance();
+            }
+            catch (ReflectiveOperationException e) {
+                throw new PrestoException(COMPILER_ERROR, e);
+            }
+        };
+    }
+
+    private ClassDefinition defineProjectionClass(RowExpression projection, InputChannels inputChannels, CallSiteBinder callSiteBinder)
+    {
+        ClassDefinition classDefinition = new ClassDefinition(
+                a(PUBLIC, FINAL),
+                makeClassName(PageProjection.class.getSimpleName()),
+                type(Object.class),
+                type(PageProjection.class));
+
+        FieldDefinition blockBuilderField = classDefinition.declareField(a(PRIVATE, FINAL), "blockBuilder", BlockBuilder.class);
+
+        CachedInstanceBinder cachedInstanceBinder = new CachedInstanceBinder(classDefinition, callSiteBinder);
+        generatePageProjectMethod(classDefinition, blockBuilderField);
+        generateProjectMethod(classDefinition, callSiteBinder, cachedInstanceBinder, projection, blockBuilderField);
+
+        // getType
+        BytecodeExpression type = invoke(callSiteBinder.bind(projection.getType(), Type.class), "type");
+        classDefinition.declareMethod(a(PUBLIC), "getType", type(Type.class))
+                .getBody()
+                .append(type.ret());
+
+        // isDeterministic
+        classDefinition.declareMethod(a(PUBLIC), "isDeterministic", type(boolean.class))
+                .getBody()
+                .append(constantBoolean(determinismEvaluator.isDeterministic(projection)).ret());
+
+        // getInputChannels
+        classDefinition.declareMethod(a(PUBLIC), "getInputChannels", type(InputChannels.class))
+                .getBody()
+                .append(invoke(callSiteBinder.bind(inputChannels, InputChannels.class), "getInputChannels").ret());
+
+        // toString
+        String toStringResult = toStringHelper(classDefinition.getType()
+                .getJavaClassName())
+                .add("projection", projection)
+                .toString();
+        classDefinition.declareMethod(a(PUBLIC), "toString", type(String.class))
+                .getBody()
+                // bind constant via invokedynamic to avoid constant pool issues due to large strings
+                .append(invoke(callSiteBinder.bind(toStringResult, String.class), "toString").ret());
+
+        // constructor
+        generateConstructor(classDefinition, cachedInstanceBinder, method -> {
+            Variable thisVariable = method.getThis();
+            BytecodeBlock body = method.getBody();
+            body.append(thisVariable.setField(
+                    blockBuilderField,
+                    type.invoke("createBlockBuilder", BlockBuilder.class, newInstance(BlockBuilderStatus.class), constantInt(1))));
```

Comment:
@dain does it make sense to reuse block builder since you have `selectedPositions` and could obtain it's length (actual number of selected positions)? If you use block builder then after `reset()` you will end up overallocating by 20% even if you filled block builder to the expected `maxBlockSizeInBytes` or `batchSize`. You will end up with empty unused (but accounted) space or with unstable `batchSize` (since after `reset()` and with the same batch size you won't meet expected block size). Please correct me if I am wrong.

Since the code is merged this can be next thing to improve. I could refactor it since it would be interesting thing to do if you don't mind.